src/telemetry-data/data-model.md

```diff
@@ -0,0 +1,203 @@
+  * while thin-edge expect all the data on a single topic.
+* Writing a mapper that groups messages as those produced by `collectd` is a more complex task than expected.
+  * One should leverage the effort done for `collectd` and use it 
+* All the data published on `tedge/measurements` are sent to the cloud.
+  * What if we wish to filter some data out?
+  * What if we wish to pre-process the data to only forward an aggregate?
+* The meaning and purpose of multi-measurements is not clear.
+* Possible name collisions with reserved key words (`time` and `type`)?
+* Non numeric values are rejected.
+* Accept only ISO-8601 timestamp, while Unix Epoch is often used by data sources. 
+* The child device identifiers are expected to be in the topic name,
+  * while some data sources published their identifier along the message.
+* There is no way to attach metadata to a data source (e.g the unit).
+* There is no way to attach metadata to a data point (e.g the quality of the measurement).
+
+## Requirements for thin-edge JSON version 1.0
+
+* Provide a clear semantics of the data that can be moved around thin-edge components.
+  * Given two sets of messages, are they equivalent or not?
+* Provide a reference mapper that move data around topics from different producers and consumers.
+* Prefer convention to configuration
+  * By convention, messages are exchanged over pre-defined topics - with a format defined by the topic.
+  * Configuration is required only to overcome a convention.
+    Ingesting data from `collectd`, should be as easy as telling thin-edge
+    to ingest not only `tedge/measurements` but also `collectd/raspberrypi`.
+* Ingest data from the main device as well as from child devices / named components.
+* Provide clear equivalence between messages independently of their shape. For instance, between:
+  * `{ "memory": { "used": 1085927424 } }` published on `tedge/measurements`
+  * `{ "memory/used": 1085927424 }` published on `tedge/measurements`
+  * `1085927424` published on `some/input/memory/used`
+* Accept ISO-8601 and Unix Epoch timestamps
+* Provide time-window batching of messages
+* Take thin-edge as a reference to provide simple mapping configuration.
+  * e.g: `[collectd/raspberrypi/${measurement.group}/${measurement.name}] ${measurement.epoch}:${measurement.value}`
+* Let the data sources add any meta-data even if not processed by thin-edge.
+  * It must be okay for a sensor to provide a serial number along its measurements.
+  * It must be okay for an OPCUA data source to tag the measurements with a quality code. 
+
+__Non goals__:
+* No complex message processing done by the reference mapper.
+  * The reference mapper only decomposes the input messages into measurement constituents
+    and recomposes output messages from these constituents.
+  * Any value transformations is delegated to external components as Apama.
+* No binary message format.
+  * The focus of thin-edge is on JSON and CSV message payload.
```

Comment:
I'm a little bit surprised that we are excluding binary payloads, as you are suggesting to use data model / schema for transposition I see not a lot of additional effort to enable binary payloads:

eg:

```plain
topic = 'binary/payload'
payload = BIN($Bytes[4]{timestamp}:$Bytes[1]{measurement.temp}:$Bytes[2]{measurement.light})
```

Reply:
This could be done, yes. But not without a substantial effort do deal with like byte ordering and a syntax to describe all possible options. This is why I prefer to deliver a first version using only clear text and see later how since can be extended.

Reply:
Of course, I agree this should probably not be a hard requirement for first version but I believe it should be optional requirement or requirement for following versions.